pkg/volume/csi/csi_plugin.go

```diff
@@ -436,24 +477,31 @@ func (p *csiPlugin) NewDetacher() (volume.Detacher, error) {
 	}, nil
 }
 
+// TODO change CanAttach to return error to propagate ability
+// to support Attachment or an error - see https://github.com/kubernetes/kubernetes/issues/74810
 func (p *csiPlugin) CanAttach(spec *volume.Spec) bool {
-	if spec.PersistentVolume == nil {
-		klog.Error(log("plugin.CanAttach test failed, spec missing PersistentVolume"))
+	driverMode, err := p.getDriverMode(spec)
+	if err != nil {
 		return false
 	}
 
-	var driverName string
-	if spec.PersistentVolume.Spec.CSI != nil {
-		driverName = spec.PersistentVolume.Spec.CSI.Driver
-	} else {
-		klog.Error(log("plugin.CanAttach test failed, spec missing CSIPersistentVolume"))
+	if driverMode == ephemeralDriverMode {
+		klog.V(4).Info(log("driver ephemeral mode detected for spec %v", spec.Name))
 		return false
 	}
 
-	skipAttach, err := p.skipAttach(driverName)
+	pvSrc, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		klog.Error(log("plugin.CanAttach failed to get info from spec: %s", err))
+		return false
+	}
+
+	driverName := pvSrc.Driver
 
+	skipAttach, err := p.skipAttach(driverName)
 	if err != nil {
 		klog.Error(log("plugin.CanAttach error when calling plugin.skipAttach for driver %s: %s", driverName, err))
+		return false
```

Comment:
don't need this return. we can let it fall through like it originally does

Reply:
if return is removed, then `!skipAttached` is returned evaluated to `true`.